tools/wptserve/wptserve/server.py

```diff
@@ -406,20 +424,116 @@ def handle_one_request(self):
+        if not isinstance(frame, RequestReceived):
+            return False
+
+        method = extract_method_header(frame.headers)
+        if method != "CONNECT":
+            return False
+
+        protocol = ""
+        for key, value in frame.headers:
+            if key in (b':protocol', u':protocol'):
+                protocol = isomorphic_encode(value)
+                break
+        if protocol != "websocket":
+            raise ProtocolError("Invalid protocol %s with CONNECT METHOD" % (protocol,))
+
+        return True
+
     def start_stream_thread(self, frame, queue):
         """
         This starts a new thread to handle frames for a specific stream.
         :param frame: The first frame on the stream
         :param queue: A queue object that the thread will use to check for new frames
         :return: The thread object that has already been started
         """
+        if self._is_extended_connect_frame(frame):
+            target = Http2WebTestRequestHandler._stream_ws_thread
+        else:
+            target = Http2WebTestRequestHandler._stream_thread
         t = threading.Thread(
-            target=Http2WebTestRequestHandler._stream_thread,
+            target=target,
             args=(self, frame.stream_id, queue)
         )
         t.start()
         return t
 
+    def _stream_ws_thread(self, stream_id, queue):
+        frame = queue.get(True, None)
+
+        rfile, wfile = os.pipe()
+        rfile, wfile = os.fdopen(rfile, 'rb'), os.fdopen(wfile, 'wb', 0)  # needs to be unbuffer for websockets
+        stream_handler = H2HandlerCopy(self, frame, rfile)
+
+        h2request = H2Request(stream_handler)
+        h2response = H2Response(stream_handler, h2request)
+
+        dispatcher = dispatch.Dispatcher(self.server.ws_doc_root, None, False)
+        if not dispatcher.get_handler_suite(stream_handler.path):
+            h2response.set_error(404)
+            h2response.write()
+            return
+
+        request_wrapper = _WebSocketRequest(stream_handler, h2response)
+
+        handshaker = WsH2Handshaker(request_wrapper, dispatcher)
+        try:
+            handshaker.do_handshake()
+        except HandshakeException as e:
+            self.logger.info('Handshake failed for error: %s', e)
+            h2response.set_error(e.status)
+            h2response.write()
+            return
+
+        # h2 Handshaker prepares the headers but does not send them down the
+        # wire. Flush the headers here.
+        h2response.write_status_headers()
+
+        request_wrapper._dispatcher = dispatcher
+
+        # TODO(arenevier): we need two threads:
+        # - one to handle the frame queue
+        # - one to handle the request (dispatcher.transfer_data is blocking)
+        # the alternative is to have only one (blocking) thread. That thread
+        # will call transfer_data. That would require a special case in
+        # handle_one_request, to bypass the queue and write data to wfile
+        # directly.
```

Comment:
It's not clear to me what this `TODO` is for. Is this something that should be fixed in the future (e.g. move to one thread)?